feat: introduced basic type map

ffMathy opened the pull request:
This just adds types so that we can modify the `ftrack-ts-schema-generator` accordingly to extend those types.

See https://github.com/ftrackhq/ftrack-ts-schema-generator/issues/11.

When this is merged, we should be able to do this in the `ftrack-ts-schema-generator` project:

```typescript
declare module "@ftrack/api" {
    //put generated types here
}
```

That should allow them to be extended.

<!--
  [For internal use] 
  Copy the task id from the bottom of the sidebar and paste it after FTRACK-

  Resolves FTRACK-

-->

- [x] I have added automatic tests where applicable
- [x] The PR title is suitable as a release note
- [x] The PR contains a description of what has been changed
- [x] The description contains manual test instructions

## Changes

<!--
  What are you changing and what is the reason? If there are any UI changes, include a screenshot for the changes.
-->

## Test

<!--
  How should this be tested? Include any requirements on other repositories or environment.
  For bug fixes, include the original reproduction steps.
-->

ffMathy committed:
feat: introduced basic type map

For later extensibility.

ffMathy committed:
feat: use stronger types

ffMathy commented:
I applied some strong-typing now as well to the queries.

ffMathy committed:
fix: tests pass

ffMathy committed:
fix: compile issues

ffMathy committed:
fix: better `ensure` types

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

ffMathy commented:
@gismya I did a lot of PRs and other things during the weekend, so just wanted to make sure that this didn't get lost in the spam you must have received.

Looking forward to your input on this one (when you have time of course!)

gismya commented on source/types.ts:
```diff
@@ -225,3 +225,11 @@ export interface QueryOptions {
+export interface EntityTypeMap {
+  FileComponent: {};
+}
+export type EntityType = keyof EntityTypeMap;
```
Setting `EntityType` like this unfortunately makes it unable to be extended - it won't automatically catch the augmented types. I'm not even sure if it works if we straight up use `keyof EntityTypeMap` all the time. We have to investigate that.

ffMathy replied:
Aha, interesting. I did not know that. That's weird to me. I'll try to think of other ways to extend the types.

ffMathy replied:
I have now switched to `keyof` everywhere. If you get a chance, it'd be awesome to test it out. If you don't have much time, I'll see if I can try it soon too!

gismya replied:
Probably won't have time in the coming two weeks. Might sneak in an extra evening session, but don't expect a review before early October unfortunately.

ffMathy replied:
Sure thing! I'll see what I can do!

gismya commented on source/types.ts:
```diff
@@ -225,3 +225,11 @@ export interface QueryOptions {
+  FileComponent: {};
+}
+export type EntityType = keyof EntityTypeMap;
+export type EntityData<TEntityType extends EntityType = EntityType> =
```
Don't add Entity here, especially not with `&` (I assume you where meaning to use `|`). Just remove Entity entirely.

ffMathy replied:
The `&` means that it has properties of both, right? The `|` would mean it has *either* of the two, which would be incorrect.

As for `Entity`, doesn't every `EntityData` have an ID and an entity type property? I thought that'd be nice to include.

ffMathy replied:
Demonstration of intersection types that have both properties:

[code/JYOwLgpgTgZghgYwgAgKJSgeygCTiAEwBtQBzZAbwChlkBnAVwSTroC5kAjTTIifANw1k0LFAD8HCsgC2EVnFIQOdMFDLIAvkM1UqoSLEQoAglDAB3bAGs6AEThg4lYXHNWotqcjDAwfFTUNTQBtAF0dPQNoeCRkM19Ve0dnalo3RLB2SmQQODlA9RByUIiqXSoAekrkABUAC3kUaKMWZDcUBEwZAAdMOggCH0xkergANwgqmq6QOmBVCHARDGxR-GIyABp2wh9G4ChkTAsQZAIUgDo9MABPHtN3GzoAJXk+uZQAXninz2SnMgAGRoVa4DYkYpCO4PX6ZV7vTCfZA-BILLIOQEg9BiPCESGkIRUWaqdb4iBopJvOgfAYo5AACigiM+HEpWWptIgAEoUQA+Fy0YAwRnMmlIgaXUTYXlpWjIEm8CBSsFMlmS6VQS5yBRKblCeXIZlgBhQEAGrR6WiKviXIiYUhq8WfS4ZdF0fXlARAA](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgKJSgeygCTiAEwBtQBzZAbwChlkBnAVwSTroC5kAjTTIifANw1k0LFAD8HCsgC2EVnFIQOdMFDLIAvkM1UqoSLEQoAglDAB3bAGs6AEThg4lYXHNWotqcjDAwfFTUNTQBtAF0dPQNoeCRkM19Ve0dnalo3RLB2SmQQODlA9RByUIiqXSoAekrkABUAC3kUaKMWZDcUBEwZAAdMOggCH0xkergANwgqmq6QOmBVCHARDGxR-GIyABp2wh9G4ChkTAsQZAIUgDo9MABPHtN3GzoAJXk+uZQAXninz2SnMgAGRoVa4DYkYpCO4PX6ZV7vTCfZA-BILLIOQEg9BiPCESGkIRUWaqdb4iBopJvOgfAYo5AACigiM+HEpWWptIgAEoUQA+Fy0YAwRnMmlIgaXUTYXlpWjIEm8CBSsFMlmS6VQS5yBRKblCeXIZlgBhQEAGrR6WiKviXIiYUhq8WfS4ZdF0fXlARAA)

gismya replied:
If I recall correctly there unfortunately are entity types that don't have an ID and an entity type property.

ffMathy replied:
I see. I removed it now! 👍

gismya commented on source/session.ts:
```diff
@@ -666,28 +668,30 @@ export class Session {
    *   Return update or create promise.
    */
 
-  ensure<T extends Data = Data>(
```
I'd rather take this separately if we can't do it properly right now. Don't like creating a `const` just to cast the data as `any`.

ffMathy replied:
Alright, noted. Will correct this soon 👍

ffMathy replied:
I have corrected it! 😄

gismya reviewed (changes requested):
A lot of it works very well - plugged it in to our repository (Where we have the generated typings setup), and after adding the type augmentations most if it worked. But there are some things that need to be addressed.

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

ffMathy committed:
fix: keyof everywhere

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

ffMathy commented:
It's working @gismya! Nice suggestion to use `keyof` everywhere.

Here's what I did.

1. I ran `npm link` in the `ftrack-javascript` project on the branch of this PR.
2. I made a new project with `npm init`.
3. I added TypeScript to the project with `tsc --init`.
4. I linked to the ftrack-javascript by doing `npm link @ftrack/api` in the new project.
5. Then I added the following code:

```typescript
import {Session} from "@ftrack/api";

interface Foo {
    foo: string,
    bar: string
}

declare module "@ftrack/api" {
    //put generated types here
    interface EntityTypeMap {
        Foo: Foo
    }
}

const session: Session = null!;
const result = session
    .query("")
    .then(r => r.data[0])
    .then(r => 'foo' in r && r.foo);
```

The resulting return type of "result" is `Promise<string | false>`, which proves that it's working.

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

gismya commented:
I have started looking at it, but since it affects almost all the use of the API in our entire application you can imagine it takes a while. This also requires a generated version of the types to work fully. Is there a way to fall back to the previous behavior if that is not the case? Otherwise, this change will break all projects when released, and even if it's for good reasons I think we need to be a bit more careful regarding that.

ffMathy commented:
Hmmm, I was under the impression that it would perhaps be compatible automatically with types not generated. Is that not the case? Does it error out?

gismya commented:
It errors out. I'm looking into how much of that if the fact that we have already started lightly typing the queries in our project and how much of that is independent of that.

gismya commented on source/session.ts:
```diff
@@ -669,28 +670,33 @@ export class Session {
 
-  ensure<T extends Data = Data>(
+  ensure<
+    T extends
+      EntityTypeMap[keyof EntityTypeMap] = EntityTypeMap[keyof EntityTypeMap],
+  >(
     entityType: string,
     data: T,
     identifyingKeys: Array<keyof T> = [],
   ): Promise<T & Entity> {
```
I'm thinking if we should make even bigger changes to the generics, in that we start using the entityType as the base of all the generics, to be able to automatically infer it from the entityType used.
So something like
```suggestion
    K extends
      keyof EntityTypeMap = keyof EntityTypeMap,
  >(
    entityType: K,
    data: EntityTypeMap[K],
    identifyingKeys: Array<keyof EntityTypeMap[K]> = [],
  ): Promise<EntityTypeMap[K] & Entity> {
```

It will make the typing stronger but will require more work in changing for people who have already started typing the calls in their applications.

What do you think @ffMathy, @jimmycallin, @lucaas?

ffMathy replied:
I love it! Nice work!

ffMathy replied:
Waiting for @jimmycallin and @lucaas comments before making the change.

ffMathy replied:
Friendly poke (cc @jimmycallin and @lucaas) 🙏 ❤️

gismya replied:
Decided against this after internal discussions

gismya replied:
We've been discussing it at the office and it's moving along. There are some things to figure out, we have some new query types that need to fit with this somehow and I need to find the time to do some testing and changes concerning that.

ffMathy replied:
Great - keep me posted. Thanks for the update!

gismya commented on source/session.ts:
```diff
@@ -669,28 +670,33 @@ export class Session {
   ): Promise<T & Entity> {
     let keys = identifyingKeys as string[];
 
+    const anyData = data as any;
```
I don't like this, but looking at it I can see that it gets a bit thorny to type the function better. We can add that as a later thing.

ffMathy replied:
Yeah that was exactly my thoughts too. It's definitely possible to strong-type it, but sometimes you have to go two steps forward and one step back I guess.

gismya reviewed (dismissed):
Once I got it up and running it was working well. 
We need to create a PR in the generate repo the make the needed changes there (I can create that unless you want to do it), and it need some updates to the documentation outlining how to connect with the generated types (And that you can return to previous behaviour by setting the typemap as { [key: string]: any } in the project. Overall a fantastic job and a great addition. 

I also added a discussion point whether we should take the opportunity to change what the generic reflects to get more properties typed.

I apologize for the delay in reviewing, assigning

ffMathy committed:
Update source/types.ts

Co-authored-by: Lars Johansson <gismya@gmail.com>

ffMathy commented:
> Once I got it up and running it was working well. We need to create a PR in the generate repo the make the needed changes there (I can create that unless you want to do it), and it need some updates to the documentation outlining how to connect with the generated types (And that you can return to previous behaviour by setting the typemap as { [key: string]: any } in the project. Overall a fantastic job and a great addition.
> 
> I also added a discussion point whether we should take the opportunity to change what the generic reflects to get more properties typed.
> 
> I apologize for the delay in reviewing, assigning

Thank you so much for this, and no worries about the delay! Very understandable.

You're very welcome to make the PR. I feel like the generator repo is still a bit hard to contribute to. If it was using the TS SDK to emit code, it'd be easier to maintain, and it'd be easy for me. It's clear it started as a small prototype and evolved from there, but could use an upgrade over time (as you also pointed out).

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

gismya committed:
Refinements

jimmycallin commented on source/session.ts:
```diff
@@ -670,14 +670,11 @@ export class Session {
-    data: T,
-    identifyingKeys: Array<keyof T> = [],
-  ): Promise<T & Entity> {
+  ensure<K extends EntityType = EntityType>(
```
I prefer more descriptive names - we use `TEntityType` internally

gismya committed:
Make the entitytype key the base for the responses as well

gismya committed:
K to TEntityType

gismya commented:
Okay, I've finally had time to refine it, it took a while to work out all the kinks. I changed the generic from being the type to the key of the type in the map. I added an ExtendibleEntityTypeMap which is the interface that we use module augmentation on, and if there's no augmentation done we default to the previous behavior of 
```
{
    [key: string]: any;
  }
```
For all data. This can be changed later on, but for now, I don't want users to be forced to generate the types to be able to use the API.

ffMathy commented:
Makes sense. Great work! Does that mean it can be merged now? Or does it need more changes? 😅

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

ffMathy committed:
Merge branch 'main' into feature/type-extensibility

ffMathy commented:
@gismya looks like @jimmycallin actually left some comments here. Unsure if you want to fix them or I should do it?

gismya commented:
They were fixed in my latest commit. Things are just a bit busy, it's on his radar :)

ffMathy commented:
Oh, didn't see. My bad!

No rush at all. I hope you're all well ❤️

jimmycallin commented:
Based on the changes in the schema repo, I think this needs to be updated to wrap things in `Partial` right?

ffMathy commented:
Yeah somehow I can't help but feel it's a separate change request IMO.

If this was a monorepo, it would have been more natural.

But since it is already an issue now, this feature technically doesn't have anything to do with it, and I'd rather fix it after merging this one.

Let me know if you disagree.

How's it going by the way? Can we get it merged? :heart:

jimmycallin commented:
ok we'll treat this as using v1 of the schema generator and will have a separate pr to make it support v2. will look at it now

jimmycallin commented on source/types.ts:
```diff
@@ -25,43 +25,39 @@ export interface CreateComponentOptions {
 
-export interface CreateResponse<T = Data> {
-  data: T;
+export interface CreateResponse<K extends EntityType = EntityType> {
```
For consistency, change this to `TEntityType` as well

gismya replied:
I prefer the distinction between string keys and types, I feel it makes it easier to understand what kind of generic is expected. But we should be consistent, how about changing them all to KEntityType?

jimmycallin replied:
didn't know that was a convention. it's new to me so i don't like it ;)

TEntityType > KEntityType > K

both of TEntityType and KEntityType are acceptable!

gismya replied:
@ffMathy - want to be the deciding vote between TEntityType and KEntityType?

ffMathy replied:
I have no preference, sorry 🤣

One sec, let me throw a coin...

Okay, it's TEntityType 🙏👌

jimmycallin commented on source/types.ts:
```diff
@@ -25,43 +25,39 @@ export interface CreateComponentOptions {
-export interface QueryResponse<T = Data> {
-  data: T[];
+export interface QueryResponse<K extends EntityType = EntityType> {
+  data: EntityData<K>[]; // Later this needs to take multiple different K and give the correct types
```
can you expand on what this comment means and how it should be resolved?

ffMathy replied:
I didn't add this actually. I think @gismya could chime in?

gismya replied:
A call can contain multiple query operations with different entity types.. But writing this I realize that would be handled in the tuple of the call and not here. I'll remove the comment.

jimmycallin commented on source/types.ts:
```diff
@@ -229,3 +225,22 @@ export interface QueryOptions {
+export type EntityData<TEntityType extends EntityType = EntityType> =
+  EntityTypeMap[TEntityType];
+
+export interface TypedContextSubtypeMap {}
```
are we using all of these types still? don't quite understand their usages in that case, and could need some comments

jimmycallin commented on source/uploader.ts:
```diff
@@ -98,11 +98,11 @@ export class Uploader {
   uploadMetadata: GetUploadMetadataResponse | null;
   /** @deprecated - Remove once Session.createComponent signature is updated. */
-  createComponentLocationResponse: CreateResponse<Data> | null;
+  createComponentLocationResponse: CreateResponse<EntityType> | null;
```
should these be `CreateResponse<"Component">` and `CreateResponse<"ComponentLocation">`?

ffMathy replied:
No idea. CC @gismya.

gismya replied:
Sounds correct. Changing.

jimmycallin commented on source/session.ts:
```diff
@@ -841,12 +844,12 @@ export class Session {
    * containing action, data and metadata
    */
-  async query<T extends Data = Data>(
+  async query<TEntityType extends EntityType>(
```
missing default value? `TEntityType extends EntityType = EntityType`

jimmycallin commented on source/operation.ts:
```diff
@@ -137,7 +139,7 @@ export function search({
+  type: EntityType,
   keys: string[] | string,
   data: any,
 ): UpdateOperation {
```
we could pass `TEntityType` as a generic to UpdateOperation, and have it be automatically inferred from the `type` param. that could be useful in `session.update` i believe, as it then gets back an update operation with the entity type properly typed to its own TEntityType. 

(same goes for all operations)

gismya replied:
That is the way to go and something I considered. However, I felt that expanding the scope of this PR wasn't the best option. 
There are quite a few different places where we can be more specific in the typing when this is implemented, but I feel that is best handled as a focused effort after this is merged.

ffMathy replied:
Alright. With that in mind, I'll mark this comment as resolved.

jimmycallin reviewed (changes requested):
great work! i think the main thing this is lacking is some documentation explaining how to properly use this, so would begin with adding that to the README. then have a look at the rest of my comments.

gismya committed:
Review Fixes

gismya committed:
Merge branch 'main' into feature/type-extensibility

gismya committed:
TS/Prettier/Import fixes

ffMathy commented:
> great work! i think the main thing this is lacking is some documentation explaining how to properly use this, so would begin with adding that to the README. then have a look at the rest of my comments.

Hmmm, but isn't the behavior for the consumer of the client the same as before? They won't notice a difference I suppose?

gismya commented:
> > great work! i think the main thing this is lacking is some documentation explaining how to properly use this, so would begin with adding that to the README. then have a look at the rest of my comments.
> 
> Hmmm, but isn't the behavior for the consumer of the client the same as before? They won't notice a difference I suppose?

If we don't document how developers can connect their generated types to the API we'll probably never remember to / take the time to document it.

But I can have a go at writing that.

gismya committed:
Made EntityType generic naming consistent

gismya commented:
I'm realizing that to write the documentation for how to use it I need to make the necessary changes to the output of `ftrack-ts-schema-generator` first. I'll try to get that and the documentation done tomorrow or Thursday.

ffMathy commented:
All of this really points to a monorepo making sense, perhaps with NX. These two projects are highly coupled.

gismya commented:
They weren't until now :)

ffMathy commented:
Yes, now they are coupled in code/syntax as well, but semantically they have always been couple. I've had numerous occasions where I needed to wait for something to be done in one repo before I could push to the other. That's also coupling.

The point remains the same, that I think the repo could benefit from a monorepo structure. But that's just my 5 cents - I am not the owner of the repo :pray:

gismya committed:
Add documentation

jimmycallin commented:
We'll close this since we preferred the other implementation. Thank you for your contribution to this!
